File: simulation/OathAndCoin.Simulation/Ids/ContentId.cs
using System.Text.RegularExpressions;

namespace OathAndCoin.Simulation.Ids;

/// <summary>
/// A stable, namespaced content identifier (ADR-005) in the form
/// <c>namespace:name</c>, where both segments match
/// <c>^[a-z][a-z0-9_]*$</c>.
/// </summary>
/// <remarks>
/// The only way to obtain a valid instance is <see cref="Parse"/> or
/// <see cref="TryParse"/> — the constructor is private, so an
/// out-of-band-invalid value cannot exist in the system. The type's own
/// <c>default</c> is the one exception C# forces on every struct; it is
/// treated as "uninitialized" rather than a silently empty id, so touching
/// <see cref="Namespace"/>, <see cref="Name"/>, or <see cref="Value"/> on a
/// <c>default(ContentId)</c> throws instead of returning <c>null</c> or
/// empty text. Ordering and hashing are ordinal only: this type never
/// consults the host machine's locale (TDD §7.3).
/// </remarks>
public readonly struct ContentId : IEquatable<ContentId>, IComparable<ContentId>
{
    private const string SegmentPatternText = "^[a-z][a-z0-9_]*$";

    private static readonly Regex SegmentPattern = new(
        SegmentPatternText,
        RegexOptions.Compiled | RegexOptions.CultureInvariant);

    // null only for default(ContentId); every instance built through Parse
    // or TryParse has all three fields populated together.
    private readonly string? _namespace;
    private readonly string? _name;
    private readonly string? _value;

    private ContentId(string @namespace, string name, string value)
    {
        _namespace = @namespace;
        _name = name;
        _value = value;
    }

    /// <summary>The namespace segment, e.g. <c>"core"</c> in <c>"core:bram"</c>.</summary>
    public string Namespace => _namespace ?? throw CreateUninitializedException();

    /// <summary>The name segment, e.g. <c>"bram"</c> in <c>"core:bram"</c>.</summary>
    public string Name => _name ?? throw CreateUninitializedException();

    /// <summary>The full <c>namespace:name</c> text.</summary>
    public string Value => _value ?? throw CreateUninitializedException();

    /// <summary>
    /// Parses <paramref name="text"/> as a <see cref="ContentId"/>.
    /// </summary>
    /// <exception cref="FormatException">
    /// <paramref name="text"/> is not a valid <c>namespace:name</c> id.
    /// </exception>
    public static ContentId Parse(string? text)
    {
        if (TryParse(text, out var id))
        {
            return id;
        }

        throw new FormatException(
            $"Invalid ContentId '{text ?? "null"}'. Expected format 'namespace:name', "
            + $"where each segment matches '{SegmentPatternText}'.");
    }

    /// <summary>
    /// Attempts to parse <paramref name="text"/> as a <see cref="ContentId"/>,
    /// without throwing on malformed input.
    /// </summary>
    public static bool TryParse(string? text, out ContentId result)
    {
        result = default;

        if (string.IsNullOrEmpty(text))
        {
            return false;
        }

        var separatorIndex = text.IndexOf(':');
        if (separatorIndex < 0 || text.IndexOf(':', separatorIndex + 1) >= 0)
        {
            return false;
        }

        var namespaceSegment = text[..separatorIndex];
        var nameSegment = text[(separatorIndex + 1)..];

        if (!SegmentPattern.IsMatch(namespaceSegment) || !SegmentPattern.IsMatch(nameSegment))
        {
            return false;
        }

        result = new ContentId(namespaceSegment, nameSegment, text);
        return true;
    }

    /// <summary>
    /// Ordinal comparison of <see cref="Value"/>, never locale-dependent.
    /// </summary>
    public int CompareTo(ContentId other) => string.CompareOrdinal(_value, other._value);

    public bool Equals(ContentId other) => string.Equals(_value, other._value, StringComparison.Ordinal);

    public override bool Equals(object? obj) => obj is ContentId other && Equals(other);

    public override int GetHashCode() => _value is null ? 0 : string.GetHashCode(_value, StringComparison.Ordinal);

    public override string ToString() => Value;

    public static bool operator ==(ContentId left, ContentId right) => left.Equals(right);

    public static bool operator !=(ContentId left, ContentId right) => !left.Equals(right);

    public static bool operator <(ContentId left, ContentId right) => left.CompareTo(right) < 0;

    public static bool operator <=(ContentId left, ContentId right) => left.CompareTo(right) <= 0;

    public static bool operator >(ContentId left, ContentId right) => left.CompareTo(right) > 0;

    public static bool operator >=(ContentId left, ContentId right) => left.CompareTo(right) >= 0;

    private static InvalidOperationException CreateUninitializedException() =>
        new("ContentId is uninitialized (default(ContentId)). Obtain a value via "
            + "ContentId.Parse or ContentId.TryParse before reading it.");
}

File: packages/simulation/src/ids/content-id.ts
/**
 * A stable, namespaced content identifier (`ADR-005`) in the form
 * `namespace:name`, where both segments match `^[a-z][a-z0-9_]*$`.
 *
 * Ported from `OathAndCoin.Simulation.Ids.ContentId`, and the port is a plain
 * branded string rather than a wrapper object. That is not a shortcut — it
 * deletes a whole class of defect the C# version had to defend against by hand:
 *
 * - The C# struct needed `ContentIdJsonConverter`, attached to the type, because
 *   `System.Text.Json` otherwise wrote `{"Namespace":"core","Name":"bram",...}`
 *   and read it back as `default(ContentId)` silently. A branded string *is* its
 *   own wire format; there is no second representation to disagree with.
 * - It needed `ReadAsPropertyName`/`WriteAsPropertyName` overrides because
 *   dictionary keys travel a different code path. A string is already a valid
 *   key everywhere.
 * - It needed a `default(ContentId)` guard — three nullable fields and an
 *   "uninitialized" exception — because C# forces a zero value on every struct.
 *   TypeScript forces nothing: the only values of this type are the ones
 *   {@link parseContentId} and {@link tryParseContentId} return.
 *
 * What the brand keeps is the guarantee those defences existed for: an
 * out-of-band-invalid identifier cannot be produced by ordinary code, because
 * the type is not assignable from `string`. A deliberate cast defeats it, the
 * same way `Unsafe` defeats the C# version.
 *
 * Ordering and comparison are ordinal — UTF-16 code units, never the host's
 * locale (`TDD` §7.3). That is what JavaScript's own `<` on strings already
 * does, so {@link compareContentIds} is a comparator over it rather than a
 * reimplementation.
 */

declare const contentIdBrand: unique symbol;

export type ContentId = string & { readonly [contentIdBrand]: 'ContentId' };

/**
 * The one statement of what a segment looks like. Everything else about the
 * format is built from it — the per-segment check, the diagnostic an author
 * reads, and the whole-identifier pattern the content schemas state — so the
 * format cannot be tightened in one of those places and left alone in the others.
 */
const SEGMENT_BODY = '[a-z][a-z0-9_]*';

const SEGMENT_PATTERN_TEXT = `^${SEGMENT_BODY}$`;

const SEGMENT_PATTERN = new RegExp(SEGMENT_PATTERN_TEXT);

/**
 * The whole-identifier pattern, for consumers that validate a `namespace:name`
 * string in one step instead of splitting it — the Zod content contracts, and
 * through them the generated JSON Schemas. Exported rather than restated there:
 * a schema whose pattern drifts from this parser is a schema that accepts
 * identifiers the loader then refuses, which is the worst of both.
 */
export const CONTENT_ID_PATTERN = `^${SEGMENT_BODY}:${SEGMENT_BODY}$`;

/**
 * Parses `text` as a {@link ContentId}.
 *
 * @throws if `text` is not a valid `namespace:name` identifier. The message is
 * the C# one word for word: a diagnostic an author has already learned to read
 * should not change spelling just because the language did.
 */
export function parseContentId(text: string | null | undefined): ContentId {
  const parsed = tryParseContentId(text);
  if (parsed === undefined) {
    throw new Error(
      `Invalid ContentId '${text ?? 'null'}'. Expected format 'namespace:name', ` +
        `where each segment matches '${SEGMENT_PATTERN_TEXT}'.`
    );
  }

  return parsed;
}

/** Parses `text`, answering `undefined` instead of throwing on malformed input. */
export function tryParseContentId(text: string | null | undefined): ContentId | undefined {
  if (text === null || text === undefined || text === '') {
    return undefined;
  }

  const separatorIndex = text.indexOf(':');
  // Exactly one separator. `a:b:c` is rejected rather than read as
  // namespace `a` and name `b:c`, which would make two different spellings of
  // one identifier both parse.
  if (separatorIndex < 0 || text.indexOf(':', separatorIndex + 1) >= 0) {
    return undefined;
  }

  const namespaceSegment = text.slice(0, separatorIndex);
  const nameSegment = text.slice(separatorIndex + 1);

  if (!SEGMENT_PATTERN.test(namespaceSegment) || !SEGMENT_PATTERN.test(nameSegment)) {
    return undefined;
  }

  return text as ContentId;
}

/** The namespace segment, e.g. `core` in `core:bram`. */
export function contentIdNamespace(id: ContentId): string {
  return id.slice(0, id.indexOf(':'));
}

/** The name segment, e.g. `bram` in `core:bram`. */
export function contentIdName(id: ContentId): string {
  return id.slice(id.indexOf(':') + 1);
}

/**
 * Ordinal comparison, never locale-dependent. Returns a negative number, zero
 * or a positive number, the shape every sort in this package expects.
 */
export function compareContentIds(left: ContentId, right: ContentId): number {
  if (left < right) {
    return -1;
  }

  return left > right ? 1 : 0;
}
